src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ConstructorParametersShouldMatchPropertyAndFieldNames.cs

```diff
@@ -0,0 +1,292 @@
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+        internal static DiagnosticDescriptor UnreferencedParameterRule = DiagnosticDescriptorHelper.Create(RuleId,
+                                                                             s_localizableTitle,
+                                                                             s_localizableMessageUnreferencedParameter,
+                                                                             DiagnosticCategory.Design,
+                                                                             RuleLevel.BuildWarning,
+                                                                             description: s_localizableDescriptionUnreferencedParameter,
+                                                                             isPortedFxCopRule: false,
+                                                                             isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(PropertyNameRule, FieldRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction((compilationStartContext) =>
+            {
+                INamedTypeSymbol? jsonConstructorAttributeNamedSymbol = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemTextJsonSerializationJsonConstructorAttribute);
+                if (jsonConstructorAttributeNamedSymbol == null)
+                {
+                    return;
+                }
+
+                var paramAnalyzer = new ParameterAnalyzer(jsonConstructorAttributeNamedSymbol);
+
+                compilationStartContext.RegisterSymbolStartAction((symbolStartContext) =>
+                {
+                    var constructors = ((INamedTypeSymbol)symbolStartContext.Symbol).InstanceConstructors;
+
+                    foreach (var ctor in constructors)
+                    {
+                        if (paramAnalyzer.ShouldAnalyzeMethod(ctor))
+                        {
+                            var referencedParameters = PooledConcurrentSet<IParameterSymbol>.GetInstance();
+
+                            symbolStartContext.RegisterOperationAction(
+                                context => ParameterAnalyzer.AnalyzeOperationAndReport(context, referencedParameters),
+                                OperationKind.ParameterReference);
+
+                            symbolStartContext.RegisterSymbolEndAction(
+                                context => ParameterAnalyzer.ReportUnusedParameters(context, ctor, referencedParameters));
```

Comment:
@buyaa-n I've updated the behaviour of the "Initialize" function accordingly with your previous comments. Could you please take a look at it? I've also added detection of unreferenced parameters (which partially duplicates the logic of `ReviewUnusedParametersAnalyzer`, and planning to also add a separate diagnostic for referenced static fields (`JsonConstructor` will throw an exception). Are you OK with it? Or is it something that shouldn't be done?

Reply:
Thanks @psxvoid

> I've also added detection of unreferenced parameters (which partially duplicates the logic of ReviewUnusedParametersAnalyzer, 

I am not sure that was necessary, the constructor parameters should match the public property/fields but checking if the parameter usage might unnecessary as the user might on purpose not setting that property, plus it might be redundant, especially when there is an existing `ReviewUnusedParametersAnalyzer` for such case

> and planning to also add a separate diagnostic for referenced static fields/props (JsonConstructor will throw an exception). Are you OK with it? Or is it something that shouldn't be done?

I think that might not necessary, by accounting developers mostly will not initiate a static field/property using constructors, especially if it adds more calculation/overhead I would prefer not to add that

Reply:
Thank you. I haven't had enough experience with Roslyn Analyzers to implement everything accordingly to the guidelines and to meet all expectations regarding this question. If it is still relevant, I'll come back to the implementation soon and prepare another non-draft pull request (to preserve the review comments of this one) rebased to the latest preview branch. There are quite a lot of issues in this pull request - mentioned and in my private to-do list :)